**Module 2 : Data Integration Module**

ezSingleCell uses 3 different methods to perform data integration namely Seurat, Harmony and fastMNN. It accepts data in two different formats namely:

1. Gene counts (in text format), or

2. Cell Ranger output

For gene counts, the test gene expression dataset comprising of 500 cells from each batch was downloaded from:

1. 10X website (https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k), or

2. Google drive link (https://drive.google.com/file/d/1SxhS1q2SgIDSwnu-4_9x4vTpriwI4Z6T/view?usp=share_link)

The metadata can also be found at the google drive link (https://drive.google.com/file/d/1Cp9bJ0_DFqh7ePr7LruDeBj3P8ojpQ-1/view?usp=share_link)

For Cellranger output, the data can be found at the google drive as below:

1. Batch 1 data (https://drive.google.com/file/d/11yc2MvzQAlSvk_BCSr9LqQjZ1lXuEwRX/view?usp=share_link)

2. Batch 2 data (https://drive.google.com/file/d/17ShgTTMx8k8IpENqKDOiPH6Fx-ns4JVX/view?usp=share_link)

![image](https://github.com/JinmiaoChenLab/ezSingleCell2/assets/8286779/74d27ad1-0d8a-4a46-9090-2843cc95f841)





